osc/osc.go

```diff
@@ -50,41 +61,43 @@ func (osc *OSC) Send(msg Message) error {
 	return err
 }
 
-func (osc *OSC) Receive(wait time.Duration) (Message, error) {
+func (osc *OSC) Receive(ctx context.Context, wait time.Duration) (Message, error) {
 	// Waits for and reads a message from Conn from net.DialUDP
-	//    wait is wait time in milliseconds
+	// wait is wait time in milliseconds
 	var (
 		msg Message
-		err error
-		wg  sync.WaitGroup
 	)
 
+	ctx, done := context.WithTimeout(ctx, wait)
+
+	// the ctx here will be use for later
+	var eg *errgroup.Group
+	eg, ctx = errgroup.WithContext(ctx)
+	defer done()
+	_ = ctx
+
 	byt := make([]byte, 8192)
 
-	fmt.Fprintf(os.Stderr, "Waiting for Reply...\n")
-	wg.Add(1) // Add one count to waitgroup
-	go func(w time.Duration, e *error) {
-		// If 5 seconds pass, print to stderr, and close waitgroup
-		time.Sleep(w)
-		*e = fmt.Errorf("Couldn't receive a response")
-		wg.Done()
-	}(wait, &err)
-	go func(e *error) {
+	osc.Debug("Waiting for Reply...")
+
+	eg.Go(func() error {
 		// Read into msg.Packet
-		//_, *e = osc.Conn.Read(msg.Packet.Bytes())
-		_, *e = osc.Conn.Read(byt)
-		wg.Done() // If read is successful, close waitgroup
-	}(&err)
-	wg.Wait()
-	wg.Add(1) // Add to the counter to guard against panicking at a negative counter
+		if _, err := osc.Conn.Read(byt); err != nil {
+			return err
+		}
+		return nil
+	})
 
-	msg.Packet.Write(byt)
+	if _, err := msg.Packet.Write(byt); err != nil {
+		return msg, err
+	}
 
-	return msg, err
+	return msg, eg.Wait()
```

Comment:
Should the eg.Wait() occur before the msg.Packet.Write(byt)?

msg.Packet is a bytes.Buffer and here, it is writing the contents of byt (of type []byte) to msg.Packet

If the eg.Wait() is not called before msg.Packet.Write(), byt will still be empty.

Reply:
I'm still not 100% on what you want to happen here at the Receive func. I need to play with it more.